File: python/problem_80.py
from typing import Optional, Tuple


class TreeNode80:
    def __init__(self, val, left=None, right=None):
        self.val = val; self.left = left; self.right = right


def deepest_node(root: Optional[TreeNode80]) -> Tuple[str, int]:
    if root is None:
        return ('', 0)
    lv, ld = deepest_node(root.left)
    rv, rd = deepest_node(root.right)
    if ld > rd: return (lv, ld + 1)
    if rd > ld: return (rv, rd + 1)
    return (root.val, 1)


def main():
    print("Problem 80:")
    root = TreeNode80('a',
        TreeNode80('b', TreeNode80('d')),
        TreeNode80('c')
    )
    assert deepest_node(root)[0] == 'd'
    print("All Tests Passed")

File: rust/problem_80.rs
pub struct TreeNode80 {
    pub val: &'static str,
    pub left: Option<Box<TreeNode80>>,
    pub right: Option<Box<TreeNode80>>,
}

pub fn deepest_node(root: &Option<Box<TreeNode80>>) -> (&'static str, usize) {
    match root {
        None => ("", 0),
        Some(n) => {
            let (lv, ld) = deepest_node(&n.left);
            let (rv, rd) = deepest_node(&n.right);
            if ld > rd { (lv, ld+1) } else if rd > ld { (rv, rd+1) } else { (n.val, 1) }
        }
    }
}

pub fn problem80() {
    println!("Problem 80:");
    let root = Some(Box::new(TreeNode80 { val:"a",
        left: Some(Box::new(TreeNode80 { val:"b",
            left: Some(Box::new(TreeNode80 { val:"d", left:None, right:None })), right:None })),
        right: Some(Box::new(TreeNode80 { val:"c", left:None, right:None })),
    }));
    assert_eq!(deepest_node(&root).0, "d");
    println!("All Tests Passed");
}
